tests_extractor.d

```diff
@@ -59,19 +59,25 @@ class TestVisitor : ASTVisitor
     override void visit(const Declaration decl)
     {
         if (decl.unittest_ !is null && hasDdocHeader(sourceCode, decl))
-            print(decl.unittest_);
+            print(decl.unittest_, decl.attributes);
 
         decl.accept(this);
     }
 
 private:
-    void print(const Unittest u)
+    void print(const Unittest u, const(Attribute)[] attributes)
     {
+        import dparse.formatter : format;
 
         // write the origin source code line
         outFile.writefln("// Line %d", u.line);
 
-        // write the unittest block
+        // write the unittest block with its attributes
+        foreach (attrib; attributes)
+        {
+            format(outFile.lockingTextWriter, attrib);
```

Comment:
Drop `lockingTextWriter` here. `lockingTextWriter` is used for std i/o by multiple threads in the same program. Since the program is single-threaded, here it does nothing than slowing it down. Also even if multiple threads were writing to the same `File`, it still wouldn't help as in all the other places you're not acquiring the same lock.

Reply:
Hehe, a `std.stdio.File` isn't a correct output range/sink, hence the need for `lockingTextWriter`. Also note that `write/writeln` uses `lockingTextWriter` under the hood: https://github.com/dlang/phobos/blob/master/std/stdio.d#L1399

I know it's not ideal, but considering that performance doesn't critical here I wanted to keep the diff small.

Reply:
> Hehe, a std.stdio.File isn't a correct output range/sink

No I mean that you can simply the code down to this:

```D
outFile.writef!`%(%s %) unittest\n{\n`(attributes);
```
> Also note that write/writeln uses lockingTextWriter under the hood

I know, I meant lockingTextWriter is used there because stdin and stdout are global variables which need to be protected, when used from multiple threads. However `lockingText(Reader|Writer)` is not used for `File`s in general, because they're more likely to br used only by a single thread, and this particular case we're definitely in a single threaded environment.